File: python3/Tries/find_all_words_on_a_board.py
from typing import List

class TrieNode:
    def __init__(self):
        self.children = {}
        self.word = None

def find_all_words_on_a_board(board: List[List[str]], words: List[str]) -> List[str]:
    root = TrieNode()
    # Insert every word into the trie.
    for word in words:
        node = root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.word = word
    res = []
    # Start a DFS call from each cell of the board that contains a
    # child of the root node, which represents the first letter of a
    # word in the trie.
    for r in range(len(board)):
        for c in range(len(board[0])):
            if board[r][c] in root.children:
                dfs(board, r, c, root.children[board[r][c]], res)
    return res

def dfs(board: List[List[str]], r: int, c: int, node: TrieNode, res: List[str]) -> None:
    # If the current node represents the end of a word, add the word to
    # the result.
    if node.word:
        res.append(node.word)
        # Ensure the current word is only added once.
        node.word = None
    temp = board[r][c]
    # Mark the current cell as visited.
    board[r][c] = '#'
    # Explore all adjacent cells that correspond with a child of the
    # current TrieNode.
    dirs = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    for d in dirs:
        next_r, next_c = r + d[0], c + d[1]
        if (is_within_bounds(next_r, next_c, board) and board[next_r][next_c] in node.children):
            dfs(board, next_r, next_c, node.children[board[next_r][next_c]], res)
    # Backtrack by reverting the cell back to its original character.
    board[r][c] = temp

def is_within_bounds(r: int, c: int, board: List[str]) -> bool:
    return 0 <= r < len(board) and 0 <= c < len(board[0])

File: kotlin/Tries/FindAllWordsOnABoard.kt
/*
    Definition of a TrieNode:

    data class TrieNode(
        var isWord: Boolean = false,
        var word: String? = null
    )
*/

fun findAllWordsOnABoard(board: Array<CharArray>, words: List<String>): List<String> {
    val root = TrieNode()
    // Insert every word into the trie.
    for (word in words) {
        var node = root
        for (c in word.toCharArray()) {
            if (c !in node.children) {
                node.children[c] = TrieNode()
            }
            node = node.children[c]!!
        }
        node.word = word
    }
    val res = mutableListOf<String>()
    // Start a DFS call from each cell of the board that contains a
    // child of the root node, which represents the first letter of a
    // word in the trie.
    for (r in 0 until board.size) {
        for (c in 0 until board[0].size) {
            if (board[r][c] in root.children) {
                dfs(board, r, c, root.children[board[r][c]], res)
            }
        }
    }
    return res
}

fun dfs(board: Array<CharArray>, r: Int, c: Int, node: TrieNode?, res: MutableList<String>) {
    // If the current node represents the end of a word, add the word to
    // the result.
    if (node?.word != null) {
        res.add(node.word!!)
        // Ensure the current word is only added once.
        node.word = null
    }
    val temp = board[r][c]
    // Mark the current cell as visited.
    board[r][c] = '#'
    // Explore all adjacent cells that correspond with a child of the
    // current TrieNode.
    val dirs = listOf(-1 to 0, 1 to 0, 0 to -1, 0 to 1)
    for ((dr, dc) in dirs) {
        val nextR = r + dr
        val nextC = c + dc
        if (isWithinBounds(board, nextR, nextC) && board[nextR][nextC] in node?.children.orEmpty()) {
            dfs(board, nextR, nextC, node?.children?.get(board[nextR][nextC]), res)
        }
        // Backtrack by reverting the cell back to its original character.
        board[r][c] = temp
    }
}

fun isWithinBounds(board: Array<CharArray>, r: Int, c: Int): Boolean {
    return r in board.indices && c in board[0].indices
}
